sql: rework the pre and post renderings for groupNode, remove identAggregator

```sql
CREATE DATABASE t;
CREATE TABLE t.t(x INT);
INSERT INTO t.t(x) VALUES (42);
SELECT 1 + count(*) FROM t.t; -- this works
TRUNCATE t.t;
SELECT 1 + count(*) FROM t.t; -- this panics!
```

```
panic: interface conversion: parser.Datum is nil, not *parser.DInt [recovered]
        panic: SELECT 1 + count(*) FROM t.t;: interface conversion: parser.Datum is nil, not *parser.DInt [recovered]
        panic: SELECT 1 + count(*) FROM t.t;: interface conversion: parser.Datum is nil, not *parser.DInt

goroutine 297 [running]:
panic(0x17da0a0, 0xc4203f5090)
        /mnt/kena/go1.7.3/src/runtime/panic.go:500 +0x1a1
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc420499dd0)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:185 +0x6e
panic(0x17da0a0, 0xc4203f5090)
        /mnt/kena/go1.7.3/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements.func1(0xc420942000, 0x1d)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:455 +0x149
panic(0x180a580, 0xc420cadac0)
        /mnt/kena/go1.7.3/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql/parser.glob..func133(0xc420aee1a8, 0x0, 0x0, 0x255ca20, 0xc4203f5068, 0x0, 0xc420211ec0, 0xc421620ea0, 0x2526cf0)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/eval.go:249 +0xdc
github.com/cockroachdb/cockroach/pkg/sql/parser.(*BinaryExpr).Eval(0xc42090fc80, 0xc420aee1a8, 0xc42169bc30, 0xc4203f5040, 0x1, 0x1)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/eval.go:1858 +0x1a8
github.com/cockroachdb/cockroach/pkg/sql.(*groupNode).computeAggregates(0xc420a86c60, 0xc42090fc00, 0x0)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/group.go:403 +0x287
```

--- a/pkg/sql/group.go
+++ b/pkg/sql/group.go
@@ -757,7 +757,20 @@ func (a *aggregateFuncHolder) Eval(ctx *parser.EvalContext) (parser.Datum, error
 		found = a.create()
 	}
 
-	return found.Result(), nil
+	result := found.Result()
+
+	if result == nil {
+		if parser.IsIdentAggregate(found) {
+			// Identity functions return their argument, even if no
+			// aggregation inputs were seen.
+			return a.arg.Eval(ctx)
+		}
+		// Otherwise, we can't be here: all aggregation functions
+		// should return a valid value or DNull if there are no rows.
+		panic("aggregation function returned nil")
+	}
+
+	return result, nil
 }
 
 func (a *aggregateFuncHolder) ResolvedType() parser.Type {

--- a/pkg/sql/parser/aggregate_builtins.go
+++ b/pkg/sql/parser/aggregate_builtins.go
@@ -217,6 +217,12 @@ type identAggregate struct {
 	val Datum
 }
 
+// IsIdentAggregate returns true for identAggregate.
+func IsIdentAggregate(f AggregateFunc) bool {
+	_, ok := f.(*identAggregate)
+	return ok
+}
+
 // NewIdentAggregate returns an identAggregate (see comment on struct).
 func NewIdentAggregate() AggregateFunc {
 	return &identAggregate{}
@@ -229,6 +235,9 @@ func (a *identAggregate) Add(datum Datum) {
 
 // Result returns the value most recently passed to Add.
 func (a *identAggregate) Result() Datum {
+	// It is significant that identAggregate returns nil, and not DNull,
+	// if no result was known via Add(). See
+	// sql.(*aggregateFuncHolder).Eval() for details.
 	return a.val
 }
 

--- a/pkg/sql/testdata/aggregate
+++ b/pkg/sql/testdata/aggregate
@@ -40,6 +40,33 @@ SELECT ARRAY_AGG(1)
 ----
 {1}
 
+# Check that COALESCE using aggregate results over an empty table
+# work properly. (#12525)
+query I
+SELECT COALESCE(MAX(1), 0) FROM generate_series(1,0)
+----
+0
+
+# Same, using arithmetic on COUNT.
+query I
+SELECT 1 + COUNT(*) FROM generate_series(1,0)
+----
+1
+
+# Same, using an empty table.
+# The following test *must* occur before the first INSERT to the tables,
+# so that it can observe an empty table.
+query II
+SELECT COUNT(*), COALESCE(MAX(k), 1) FROM kv
+----
+0 1
+
+# Same, using a subquery. (#12705)
+query I
+SELECT (SELECT COALESCE(MAX(1), 0) FROM generate_series(1,0))
+----
+0
+
 statement OK
 INSERT INTO kv VALUES
 (1, 2, 3, 'a'),

